### PR TITLE
Dynamic optimization lock

### PR DESCRIPTION
**Pull request by odespard**
Reduce the size of the optimization lock as detection of precursors increases.

### CODE REVIEW

File: alphadia/workflow/optimization.py [hunk 1]
```diff
@@ -95,24 +96,42 @@ def step(
...
+                    precursors_df.sort_values(by="precursor_idx")
+                    .iloc[
+                        2
+                        * self.workflow.config["calibration"][
+                            "optimization_lock_target"
+                        ]
+                    ]
+                    .precursor_idx
+                )
+            except IndexError:
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:112]: I know that "favor 'try' over 'upfront check'" is the pythonic way, but I feel in this case it would increase readability if the check was explictly done ..
as far as I got it
```python
num_targets = 2* self.workflow.config["calibration"][
                            "optimization_lock_target"
                        ]
if len(precursors_df) >=  num_targets:
...
else
...
```

File: alphadia/workflow/optimization.py [hunk 2]
```diff
@@ -95,24 +96,42 @@ def step(
                 verbosity="progress",
             )
             return
+        if record_step:
+            try:
+                # Takes the target number of precursors if at least the target is found,
+                precursor_cutoff_idx = (
+                    precursors_df.sort_values(by="precursor_idx")
+                    .iloc[
+                        2
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:105]: sure you want to pick only one index? or is there a `:` missing ;-)

File: alphadia/workflow/optimization.py [hunk 3]
```diff
@@ -335,6 +358,7 @@ def step(
         precursors_df: pd.DataFrame,
         fragments_df: pd.DataFrame,
         current_step: int = -1,
+        record_step: bool = False,
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:361]: please add a docstring explaining this parameter

File: alphadia/workflow/optimization.py [hunk 4]
```diff
@@ -87,6 +87,7 @@ def step(
         precursors_df: pd.DataFrame,
         fragments_df: pd.DataFrame,
         current_step: int = -1,
+        record_step: bool = True,
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:90]: please add a docstring explaining this parameter

File: alphadia/workflow/peptidecentric.py [hunk 5]
```diff
@@ -366,18 +366,33 @@ def get_optimization_lock(self):
...
 
         optimization_lock_eg_idxes = self.elution_group_order[:final_stop_index]
-
         optimization_lock_library_precursor_df = self.spectral_library._precursor_df[
             self.spectral_library._precursor_df["elution_group_idx"].isin(
                 optimization_lock_eg_idxes
             )
         ]
 
+        randomized_idx = np.random.permutation(
```
Comment (mschwoer) [inline alphadia/workflow/peptidecentric.py:375]: in order to increase reproducibility: should we fix the seed? or at least enable fixing it (e.g. via config)?

File: alphadia/workflow/peptidecentric.py [hunk 6]
```diff
@@ -366,18 +366,33 @@ def get_optimization_lock(self):
...
+            optimization_lock_library_precursor_df["precursor_idx"].values
+        )
+        self.randomized_idx_mapping = dict(
+            zip(
+                optimization_lock_library_precursor_df["precursor_idx"].values,
+                randomized_idx,
+            )
```
Comment (mschwoer) [inline alphadia/workflow/peptidecentric.py:382]: please add `strict=True` to `zip`

Comment (mschwoer) [reply alphadia/workflow/peptidecentric.py:382]: (also below)

File: alphadia/workflow/peptidecentric.py [hunk 7]
```diff
@@ -537,7 +553,53 @@ def first_recalibration_and_optimization(
...
+            "precursor_idx"
+        ].map(self.randomized_idx_mapping)
+        try:
+            # Takes the target number of precursors if at least the target is found
+            self.precursor_cutoff_idx = (
+                precursor_df_filtered.sort_values(by="precursor_idx")
+                .iloc[self.config["calibration"]["optimization_lock_target"]]
+                .precursor_idx
+            )
+        except IndexError:
```
Comment (mschwoer) [inline alphadia/workflow/peptidecentric.py:570]: same as above: I would prefer an explicit check

File: alphadia/workflow/peptidecentric.py [hunk 8]
```diff
@@ -537,7 +553,53 @@ def first_recalibration_and_optimization(
...
+
+    def set_reduced_optimization_lock(self):
+        self.reduced_optimization_lock_library_precursor_df = (
+            self.optimization_lock_library_precursor_df[
+                self.optimization_lock_library_precursor_df["precursor_idx"]
+                < self.precursor_cutoff_idx
+            ]
+        )
+
+        fragment_cutoff_idx = self.reduced_optimization_lock_library_precursor_df[
```
Comment (mschwoer) [inline alphadia/workflow/peptidecentric.py:589]: this is a reeeeally long variable name ..

think about introucing `optlock` as subsitute for `optimization_lock` throughout all the codebase (if you choose to do so, please do the renaming in a dedicated PR)

then this could become `red_optlock_lib_precursor_df`

File: alphadia/workflow/optimization.py (hunk 2 above)
Comment (odespard) [reply alphadia/workflow/optimization.py:105]: In this case, we just want the correct cutoff index, so only one index is wanted.

Comment (mschwoer) [reply alphadia/workflow/optimization.py:105]: all good, maybe introduce `cutoff_index = 2 * .. ` then and add a little explanation (e.g. on the factor of 2 ;-))
